File: src/bun.js/webcore/EncodingLabel.zig
/// https://encoding.spec.whatwg.org/encodings.json
pub const EncodingLabel = enum {
    @"UTF-8",
    IBM866,
    @"ISO-8859-3",
    @"ISO-8859-6",
    @"ISO-8859-7",
    @"ISO-8859-8",
    @"ISO-8859-8-I",
    @"KOI8-U",
    @"windows-874",
    /// Also known as
    /// - ASCII
    /// - latin1
    @"windows-1252",
    @"windows-1253",
    @"windows-1255",
    @"windows-1257",
    Big5,
    @"EUC-JP",
    @"ISO-2022-JP",
    Shift_JIS,
    @"EUC-KR",
    @"UTF-16BE",
    @"UTF-16LE",
    @"x-user-defined",
    replacement,
    GBK,
    GB18030,

    pub fn getLabel(this: EncodingLabel) []const u8 {
        return switch (this) {
            .@"UTF-8" => "utf-8",
            .@"UTF-16LE" => "utf-16le",
            .@"UTF-16BE" => "utf-16be",
            .@"windows-1252" => "windows-1252",
            .IBM866 => "ibm866",
            .@"ISO-8859-3" => "iso-8859-3",
            .@"ISO-8859-6" => "iso-8859-6",
            .@"ISO-8859-7" => "iso-8859-7",
            .@"ISO-8859-8" => "iso-8859-8",
            .@"ISO-8859-8-I" => "iso-8859-8-i",
            .@"KOI8-U" => "koi8-u",
            .@"windows-874" => "windows-874",
            .@"windows-1253" => "windows-1253",
            .@"windows-1255" => "windows-1255",
            .@"windows-1257" => "windows-1257",
            .Big5 => "big5",
            .@"EUC-JP" => "euc-jp",
            .@"ISO-2022-JP" => "iso-2022-jp",
            .Shift_JIS => "shift_jis",
            .@"EUC-KR" => "euc-kr",
            .@"x-user-defined" => "x-user-defined",
            .replacement => "replacement",
            .GBK => "gbk",
            .GB18030 => "gb18030",
        };
    }

    pub const latin1 = EncodingLabel.@"windows-1252";

    const string_map = bun.ComptimeStringMap(EncodingLabel, .{
        // Windows-1252 (Latin1) aliases
        .{ "l1", latin1 },
        .{ "ascii", latin1 },
        .{ "cp819", latin1 },
        .{ "cp1252", latin1 },
        .{ "ibm819", latin1 },
        .{ "latin1", latin1 },
        .{ "iso88591", latin1 },
        .{ "us-ascii", latin1 },
        .{ "x-cp1252", latin1 },
        .{ "iso8859-1", latin1 },
        .{ "iso_8859-1", latin1 },
        .{ "iso-8859-1", latin1 },
        .{ "iso-ir-100", latin1 },
        .{ "csisolatin1", latin1 },
        .{ "windows-1252", latin1 },
        .{ "ansi_x3.4-1968", latin1 },
        .{ "iso_8859-1:1987", latin1 },

        // UTF-16LE aliases
        .{ "ucs-2", .@"UTF-16LE" },
        .{ "utf-16", .@"UTF-16LE" },
        .{ "unicode", .@"UTF-16LE" },
        .{ "utf-16le", .@"UTF-16LE" },
        .{ "csunicode", .@"UTF-16LE" },
        .{ "unicodefeff", .@"UTF-16LE" },
        .{ "iso-10646-ucs-2", .@"UTF-16LE" },

        // UTF-16BE aliases
        .{ "utf-16be", .@"UTF-16BE" },

        // UTF-8 aliases
        .{ "utf8", .@"UTF-8" },
        .{ "utf-8", .@"UTF-8" },
        .{ "unicode11utf8", .@"UTF-8" },
        .{ "unicode20utf8", .@"UTF-8" },
        .{ "x-unicode20utf8", .@"UTF-8" },
        .{ "unicode-1-1-utf-8", .@"UTF-8" },

        // IBM866 aliases
        .{ "ibm866", .IBM866 },
        .{ "cp866", .IBM866 },
        .{ "866", .IBM866 },
        .{ "csibm866", .IBM866 },

        // ISO-8859-3 aliases
        .{ "iso-8859-3", .@"ISO-8859-3" },
        .{ "iso8859-3", .@"ISO-8859-3" },
        .{ "iso_8859-3", .@"ISO-8859-3" },
        .{ "latin3", .@"ISO-8859-3" },
        .{ "csisolatin3", .@"ISO-8859-3" },
        .{ "iso-ir-109", .@"ISO-8859-3" },
        .{ "l3", .@"ISO-8859-3" },

        // ISO-8859-6 aliases
        .{ "iso-8859-6", .@"ISO-8859-6" },
        .{ "iso8859-6", .@"ISO-8859-6" },
        .{ "iso_8859-6", .@"ISO-8859-6" },
        .{ "arabic", .@"ISO-8859-6" },
        .{ "csisolatinarabic", .@"ISO-8859-6" },
        .{ "iso-ir-127", .@"ISO-8859-6" },
        .{ "asmo-708", .@"ISO-8859-6" },
        .{ "ecma-114", .@"ISO-8859-6" },

        // ISO-8859-7 aliases
        .{ "iso-8859-7", .@"ISO-8859-7" },
        .{ "iso8859-7", .@"ISO-8859-7" },
        .{ "iso_8859-7", .@"ISO-8859-7" },
        .{ "greek", .@"ISO-8859-7" },
        .{ "greek8", .@"ISO-8859-7" },
        .{ "csisolatingreek", .@"ISO-8859-7" },
        .{ "iso-ir-126", .@"ISO-8859-7" },
        .{ "ecma-118", .@"ISO-8859-7" },
        .{ "elot_928", .@"ISO-8859-7" },

        // ISO-8859-8 aliases
        .{ "iso-8859-8", .@"ISO-8859-8" },
        .{ "iso8859-8", .@"ISO-8859-8" },
        .{ "iso_8859-8", .@"ISO-8859-8" },
        .{ "hebrew", .@"ISO-8859-8" },
        .{ "csisolatinhebrew", .@"ISO-8859-8" },
        .{ "iso-ir-138", .@"ISO-8859-8" },
        .{ "visual", .@"ISO-8859-8" },

        // ISO-8859-8-I aliases
        .{ "iso-8859-8-i", .@"ISO-8859-8-I" },
        .{ "logical", .@"ISO-8859-8-I" },
        .{ "csiso88598i", .@"ISO-8859-8-I" },

        // KOI8-U aliases
        .{ "koi8-u", .@"KOI8-U" },
        .{ "koi8-ru", .@"KOI8-U" },

        // Windows code pages
        .{ "windows-874", .@"windows-874" },
        .{ "dos-874", .@"windows-874" },
        .{ "iso-8859-11", .@"windows-874" },
        .{ "iso8859-11", .@"windows-874" },
        .{ "iso885911", .@"windows-874" },
        .{ "iso_8859-11", .@"windows-874" },
        .{ "tis-620", .@"windows-874" },

        .{ "windows-1253", .@"windows-1253" },
        .{ "cp1253", .@"windows-1253" },
        .{ "x-cp1253", .@"windows-1253" },

        .{ "windows-1255", .@"windows-1255" },
        .{ "cp1255", .@"windows-1255" },
        .{ "x-cp1255", .@"windows-1255" },

        .{ "windows-1257", .@"windows-1257" },
        .{ "cp1257", .@"windows-1257" },
        .{ "x-cp1257", .@"windows-1257" },

        // CJK encodings
        .{ "big5", .Big5 },
        .{ "big5-hkscs", .Big5 },
        .{ "cn-big5", .Big5 },
        .{ "csbig5", .Big5 },
        .{ "x-x-big5", .Big5 },

        .{ "euc-jp", .@"EUC-JP" },
        .{ "cseucpkdfmtjapanese", .@"EUC-JP" },
        .{ "x-euc-jp", .@"EUC-JP" },

        .{ "iso-2022-jp", .@"ISO-2022-JP" },
        .{ "csiso2022jp", .@"ISO-2022-JP" },

        .{ "shift_jis", .Shift_JIS },
        .{ "shift-jis", .Shift_JIS },
        .{ "sjis", .Shift_JIS },
        .{ "csshiftjis", .Shift_JIS },
        .{ "ms932", .Shift_JIS },
        .{ "ms_kanji", .Shift_JIS },
        .{ "windows-31j", .Shift_JIS },
        .{ "x-sjis", .Shift_JIS },

        .{ "euc-kr", .@"EUC-KR" },
        .{ "cseuckr", .@"EUC-KR" },
        .{ "csksc56011987", .@"EUC-KR" },
        .{ "iso-ir-149", .@"EUC-KR" },
        .{ "korean", .@"EUC-KR" },
        .{ "ks_c_5601-1987", .@"EUC-KR" },
        .{ "ks_c_5601-1989", .@"EUC-KR" },
        .{ "ksc5601", .@"EUC-KR" },
        .{ "ksc_5601", .@"EUC-KR" },
        .{ "windows-949", .@"EUC-KR" },

        // Chinese encodings
        .{ "gbk", .GBK },
        .{ "gb2312", .GBK },
        .{ "chinese", .GBK },
        .{ "csgb2312", .GBK },
        .{ "csiso58gb231280", .GBK },
        .{ "gb_2312", .GBK },
        .{ "gb_2312-80", .GBK },
        .{ "iso-ir-58", .GBK },
        .{ "x-gbk", .GBK },

        .{ "gb18030", .GB18030 },

        // Other
        .{ "x-user-defined", .@"x-user-defined" },
        .{ "replacement", .replacement },
    });

    pub fn which(input_: string) ?EncodingLabel {
        const input = strings.trim(input_, " \t\r\n\x0C");
        return string_map.getAnyCase(input);
    }
};
const string = []const u8;

const encoding = @import("./encoding.zig");

const bun = @import("bun");
const strings = bun.strings;

File: src/runtime/webcore/EncodingLabel.rs
use bun_core::strings;

/// Every encoding the Encoding Standard defines, keyed by its canonical name.
/// https://encoding.spec.whatwg.org/#names-and-labels
#[derive(Copy, Clone, Eq, PartialEq, Debug)]
pub enum EncodingLabel {
    Utf8,
    Ibm866,
    Iso8859_2,
    Iso8859_3,
    Iso8859_4,
    Iso8859_5,
    Iso8859_6,
    Iso8859_7,
    Iso8859_8,
    Iso8859_8I,
    Iso8859_10,
    Iso8859_13,
    Iso8859_14,
    Iso8859_15,
    Iso8859_16,
    Koi8R,
    Koi8U,
    Macintosh,
    Windows874,
    Windows1250,
    Windows1251,
    /// Also known as
    /// - ASCII
    /// - latin1
    Windows1252,
    Windows1253,
    Windows1254,
    Windows1255,
    Windows1256,
    Windows1257,
    Windows1258,
    XMacCyrillic,
    Big5,
    EucJp,
    Iso2022Jp,
    ShiftJis,
    EucKr,
    Utf16Be,
    Utf16Le,
    XUserDefined,
    /// `TextDecoder` must reject this one at construction
    /// (https://encoding.spec.whatwg.org/#dom-textdecoder). It is a distinct
    /// variant so the constructor can tell it apart from an unknown label.
    Replacement,
    Gbk,
    Gb18030,
}

impl EncodingLabel {
    /// The canonical name, lowercased: the value `TextDecoder.prototype.encoding`
    /// reports, and the name used to look up the WebKit `TextCodec`.
    pub fn get_label(self) -> &'static [u8] {
        match self {
            Self::Utf8 => b"utf-8",
            Self::Utf16Le => b"utf-16le",
            Self::Utf16Be => b"utf-16be",
            Self::Windows1252 => b"windows-1252",
            Self::Ibm866 => b"ibm866",
            Self::Iso8859_2 => b"iso-8859-2",
            Self::Iso8859_3 => b"iso-8859-3",
            Self::Iso8859_4 => b"iso-8859-4",
            Self::Iso8859_5 => b"iso-8859-5",
            Self::Iso8859_6 => b"iso-8859-6",
            Self::Iso8859_7 => b"iso-8859-7",
            Self::Iso8859_8 => b"iso-8859-8",
            Self::Iso8859_8I => b"iso-8859-8-i",
            Self::Iso8859_10 => b"iso-8859-10",
            Self::Iso8859_13 => b"iso-8859-13",
            Self::Iso8859_14 => b"iso-8859-14",
            Self::Iso8859_15 => b"iso-8859-15",
            Self::Iso8859_16 => b"iso-8859-16",
            Self::Koi8R => b"koi8-r",
            Self::Koi8U => b"koi8-u",
            Self::Macintosh => b"macintosh",
            Self::Windows874 => b"windows-874",
            Self::Windows1250 => b"windows-1250",
            Self::Windows1251 => b"windows-1251",
            Self::Windows1253 => b"windows-1253",
            Self::Windows1254 => b"windows-1254",
            Self::Windows1255 => b"windows-1255",
            Self::Windows1256 => b"windows-1256",
            Self::Windows1257 => b"windows-1257",
            Self::Windows1258 => b"windows-1258",
            Self::XMacCyrillic => b"x-mac-cyrillic",
            Self::Big5 => b"big5",
            Self::EucJp => b"euc-jp",
            Self::Iso2022Jp => b"iso-2022-jp",
            Self::ShiftJis => b"shift_jis",
            Self::EucKr => b"euc-kr",
            Self::XUserDefined => b"x-user-defined",
            Self::Replacement => b"replacement",
            Self::Gbk => b"gbk",
            Self::Gb18030 => b"gb18030",
        }
    }

    pub const LATIN1: EncodingLabel = EncodingLabel::Windows1252;

    /// https://encoding.spec.whatwg.org/#concept-encoding-get
    pub fn which(input_: &[u8]) -> Option<EncodingLabel> {
        // ASCII whitespace: TAB, LF, FF, CR, SPACE.
        let input = strings::trim(input_, b" \t\r\n\x0C");
        strings::in_map_case_insensitive(input, &STRING_MAP)
    }
}

// The complete label table from https://encoding.spec.whatwg.org/encodings.json.
// Every label the spec defines, and only those, must resolve here.
bun_core::comptime_string_map! {
    static STRING_MAP: EncodingLabel = {
    // UTF-8
    b"unicode-1-1-utf-8" => EncodingLabel::Utf8,
    b"unicode11utf8" => EncodingLabel::Utf8,
    b"unicode20utf8" => EncodingLabel::Utf8,
    b"utf-8" => EncodingLabel::Utf8,
    b"utf8" => EncodingLabel::Utf8,
    b"x-unicode20utf8" => EncodingLabel::Utf8,

    // IBM866
    b"866" => EncodingLabel::Ibm866,
    b"cp866" => EncodingLabel::Ibm866,
    b"csibm866" => EncodingLabel::Ibm866,
    b"ibm866" => EncodingLabel::Ibm866,

    // ISO-8859-2
    b"csisolatin2" => EncodingLabel::Iso8859_2,
    b"iso-8859-2" => EncodingLabel::Iso8859_2,
    b"iso-ir-101" => EncodingLabel::Iso8859_2,
    b"iso8859-2" => EncodingLabel::Iso8859_2,
    b"iso88592" => EncodingLabel::Iso8859_2,
    b"iso_8859-2" => EncodingLabel::Iso8859_2,
    b"iso_8859-2:1987" => EncodingLabel::Iso8859_2,
    b"l2" => EncodingLabel::Iso8859_2,
    b"latin2" => EncodingLabel::Iso8859_2,

    // ISO-8859-3
    b"csisolatin3" => EncodingLabel::Iso8859_3,
    b"iso-8859-3" => EncodingLabel::Iso8859_3,
    b"iso-ir-109" => EncodingLabel::Iso8859_3,
    b"iso8859-3" => EncodingLabel::Iso8859_3,
    b"iso88593" => EncodingLabel::Iso8859_3,
    b"iso_8859-3" => EncodingLabel::Iso8859_3,
    b"iso_8859-3:1988" => EncodingLabel::Iso8859_3,
    b"l3" => EncodingLabel::Iso8859_3,
    b"latin3" => EncodingLabel::Iso8859_3,

    // ISO-8859-4
    b"csisolatin4" => EncodingLabel::Iso8859_4,
    b"iso-8859-4" => EncodingLabel::Iso8859_4,
    b"iso-ir-110" => EncodingLabel::Iso8859_4,
    b"iso8859-4" => EncodingLabel::Iso8859_4,
    b"iso88594" => EncodingLabel::Iso8859_4,
    b"iso_8859-4" => EncodingLabel::Iso8859_4,
    b"iso_8859-4:1988" => EncodingLabel::Iso8859_4,
    b"l4" => EncodingLabel::Iso8859_4,
    b"latin4" => EncodingLabel::Iso8859_4,

    // ISO-8859-5
    b"csisolatincyrillic" => EncodingLabel::Iso8859_5,
    b"cyrillic" => EncodingLabel::Iso8859_5,
    b"iso-8859-5" => EncodingLabel::Iso8859_5,
    b"iso-ir-144" => EncodingLabel::Iso8859_5,
    b"iso8859-5" => EncodingLabel::Iso8859_5,
    b"iso88595" => EncodingLabel::Iso8859_5,
    b"iso_8859-5" => EncodingLabel::Iso8859_5,
    b"iso_8859-5:1988" => EncodingLabel::Iso8859_5,

    // ISO-8859-6
    b"arabic" => EncodingLabel::Iso8859_6,
    b"asmo-708" => EncodingLabel::Iso8859_6,
    b"csiso88596e" => EncodingLabel::Iso8859_6,
    b"csiso88596i" => EncodingLabel::Iso8859_6,
    b"csisolatinarabic" => EncodingLabel::Iso8859_6,
    b"ecma-114" => EncodingLabel::Iso8859_6,
    b"iso-8859-6" => EncodingLabel::Iso8859_6,
    b"iso-8859-6-e" => EncodingLabel::Iso8859_6,
    b"iso-8859-6-i" => EncodingLabel::Iso8859_6,
    b"iso-ir-127" => EncodingLabel::Iso8859_6,
    b"iso8859-6" => EncodingLabel::Iso8859_6,
    b"iso88596" => EncodingLabel::Iso8859_6,
    b"iso_8859-6" => EncodingLabel::Iso8859_6,
    b"iso_8859-6:1987" => EncodingLabel::Iso8859_6,

    // ISO-8859-7
    b"csisolatingreek" => EncodingLabel::Iso8859_7,
    b"ecma-118" => EncodingLabel::Iso8859_7,
    b"elot_928" => EncodingLabel::Iso8859_7,
    b"greek" => EncodingLabel::Iso8859_7,
    b"greek8" => EncodingLabel::Iso8859_7,
    b"iso-8859-7" => EncodingLabel::Iso8859_7,
    b"iso-ir-126" => EncodingLabel::Iso8859_7,
    b"iso8859-7" => EncodingLabel::Iso8859_7,
    b"iso88597" => EncodingLabel::Iso8859_7,
    b"iso_8859-7" => EncodingLabel::Iso8859_7,
    b"iso_8859-7:1987" => EncodingLabel::Iso8859_7,
    b"sun_eu_greek" => EncodingLabel::Iso8859_7,

    // ISO-8859-8
    b"csiso88598e" => EncodingLabel::Iso8859_8,
    b"csisolatinhebrew" => EncodingLabel::Iso8859_8,
    b"hebrew" => EncodingLabel::Iso8859_8,
    b"iso-8859-8" => EncodingLabel::Iso8859_8,
    b"iso-8859-8-e" => EncodingLabel::Iso8859_8,
    b"iso-ir-138" => EncodingLabel::Iso8859_8,
    b"iso8859-8" => EncodingLabel::Iso8859_8,
    b"iso88598" => EncodingLabel::Iso8859_8,
    b"iso_8859-8" => EncodingLabel::Iso8859_8,
    b"iso_8859-8:1988" => EncodingLabel::Iso8859_8,
    b"visual" => EncodingLabel::Iso8859_8,

    // ISO-8859-8-I
    b"csiso88598i" => EncodingLabel::Iso8859_8I,
    b"iso-8859-8-i" => EncodingLabel::Iso8859_8I,
    b"logical" => EncodingLabel::Iso8859_8I,

    // ISO-8859-10
    b"csisolatin6" => EncodingLabel::Iso8859_10,
    b"iso-8859-10" => EncodingLabel::Iso8859_10,
    b"iso-ir-157" => EncodingLabel::Iso8859_10,
    b"iso8859-10" => EncodingLabel::Iso8859_10,
    b"iso885910" => EncodingLabel::Iso8859_10,
    b"l6" => EncodingLabel::Iso8859_10,
    b"latin6" => EncodingLabel::Iso8859_10,

    // ISO-8859-13
    b"iso-8859-13" => EncodingLabel::Iso8859_13,
    b"iso8859-13" => EncodingLabel::Iso8859_13,
    b"iso885913" => EncodingLabel::Iso8859_13,

    // ISO-8859-14
    b"iso-8859-14" => EncodingLabel::Iso8859_14,
    b"iso8859-14" => EncodingLabel::Iso8859_14,
    b"iso885914" => EncodingLabel::Iso8859_14,

    // ISO-8859-15
    b"csisolatin9" => EncodingLabel::Iso8859_15,
    b"iso-8859-15" => EncodingLabel::Iso8859_15,
    b"iso8859-15" => EncodingLabel::Iso8859_15,
    b"iso885915" => EncodingLabel::Iso8859_15,
    b"iso_8859-15" => EncodingLabel::Iso8859_15,
    b"l9" => EncodingLabel::Iso8859_15,

    // ISO-8859-16
    b"iso-8859-16" => EncodingLabel::Iso8859_16,

    // KOI8-R
    b"cskoi8r" => EncodingLabel::Koi8R,
    b"koi" => EncodingLabel::Koi8R,
    b"koi8" => EncodingLabel::Koi8R,
    b"koi8-r" => EncodingLabel::Koi8R,
    b"koi8_r" => EncodingLabel::Koi8R,

    // KOI8-U
    b"koi8-ru" => EncodingLabel::Koi8U,
    b"koi8-u" => EncodingLabel::Koi8U,

    // macintosh
    b"csmacintosh" => EncodingLabel::Macintosh,
    b"mac" => EncodingLabel::Macintosh,
    b"macintosh" => EncodingLabel::Macintosh,
    b"x-mac-roman" => EncodingLabel::Macintosh,

    // windows-874
    b"dos-874" => EncodingLabel::Windows874,
    b"iso-8859-11" => EncodingLabel::Windows874,
    b"iso8859-11" => EncodingLabel::Windows874,
    b"iso885911" => EncodingLabel::Windows874,
    b"tis-620" => EncodingLabel::Windows874,
    b"windows-874" => EncodingLabel::Windows874,

    // windows-1250
    b"cp1250" => EncodingLabel::Windows1250,
    b"windows-1250" => EncodingLabel::Windows1250,
    b"x-cp1250" => EncodingLabel::Windows1250,

    // windows-1251
    b"cp1251" => EncodingLabel::Windows1251,
    b"windows-1251" => EncodingLabel::Windows1251,
    b"x-cp1251" => EncodingLabel::Windows1251,

    // windows-1252
    b"ansi_x3.4-1968" => EncodingLabel::LATIN1,
    b"ascii" => EncodingLabel::LATIN1,
    b"cp1252" => EncodingLabel::LATIN1,
    b"cp819" => EncodingLabel::LATIN1,
    b"csisolatin1" => EncodingLabel::LATIN1,
    b"ibm819" => EncodingLabel::LATIN1,
    b"iso-8859-1" => EncodingLabel::LATIN1,
    b"iso-ir-100" => EncodingLabel::LATIN1,
    b"iso8859-1" => EncodingLabel::LATIN1,
    b"iso88591" => EncodingLabel::LATIN1,
    b"iso_8859-1" => EncodingLabel::LATIN1,
    b"iso_8859-1:1987" => EncodingLabel::LATIN1,
    b"l1" => EncodingLabel::LATIN1,
    b"latin1" => EncodingLabel::LATIN1,
    b"us-ascii" => EncodingLabel::LATIN1,
    b"windows-1252" => EncodingLabel::LATIN1,
    b"x-cp1252" => EncodingLabel::LATIN1,

    // windows-1253
    b"cp1253" => EncodingLabel::Windows1253,
    b"windows-1253" => EncodingLabel::Windows1253,
    b"x-cp1253" => EncodingLabel::Windows1253,

    // windows-1254
    b"cp1254" => EncodingLabel::Windows1254,
    b"csisolatin5" => EncodingLabel::Windows1254,
    b"iso-8859-9" => EncodingLabel::Windows1254,
    b"iso-ir-148" => EncodingLabel::Windows1254,
    b"iso8859-9" => EncodingLabel::Windows1254,
    b"iso88599" => EncodingLabel::Windows1254,
    b"iso_8859-9" => EncodingLabel::Windows1254,
    b"iso_8859-9:1989" => EncodingLabel::Windows1254,
    b"l5" => EncodingLabel::Windows1254,
    b"latin5" => EncodingLabel::Windows1254,
    b"windows-1254" => EncodingLabel::Windows1254,
    b"x-cp1254" => EncodingLabel::Windows1254,

    // windows-1255
    b"cp1255" => EncodingLabel::Windows1255,
    b"windows-1255" => EncodingLabel::Windows1255,
    b"x-cp1255" => EncodingLabel::Windows1255,

    // windows-1256
    b"cp1256" => EncodingLabel::Windows1256,
    b"windows-1256" => EncodingLabel::Windows1256,
    b"x-cp1256" => EncodingLabel::Windows1256,

    // windows-1257
    b"cp1257" => EncodingLabel::Windows1257,
    b"windows-1257" => EncodingLabel::Windows1257,
    b"x-cp1257" => EncodingLabel::Windows1257,

    // windows-1258
    b"cp1258" => EncodingLabel::Windows1258,
    b"windows-1258" => EncodingLabel::Windows1258,
    b"x-cp1258" => EncodingLabel::Windows1258,

    // x-mac-cyrillic
    b"x-mac-cyrillic" => EncodingLabel::XMacCyrillic,
    b"x-mac-ukrainian" => EncodingLabel::XMacCyrillic,

    // GBK
    b"chinese" => EncodingLabel::Gbk,
    b"csgb2312" => EncodingLabel::Gbk,
    b"csiso58gb231280" => EncodingLabel::Gbk,
    b"gb2312" => EncodingLabel::Gbk,
    b"gb_2312" => EncodingLabel::Gbk,
    b"gb_2312-80" => EncodingLabel::Gbk,
    b"gbk" => EncodingLabel::Gbk,
    b"iso-ir-58" => EncodingLabel::Gbk,
    b"x-gbk" => EncodingLabel::Gbk,

    // gb18030
    b"gb18030" => EncodingLabel::Gb18030,

    // Big5
    b"big5" => EncodingLabel::Big5,
    b"big5-hkscs" => EncodingLabel::Big5,
    b"cn-big5" => EncodingLabel::Big5,
    b"csbig5" => EncodingLabel::Big5,
    b"x-x-big5" => EncodingLabel::Big5,

    // EUC-JP
    b"cseucpkdfmtjapanese" => EncodingLabel::EucJp,
    b"euc-jp" => EncodingLabel::EucJp,
    b"x-euc-jp" => EncodingLabel::EucJp,

    // ISO-2022-JP
    b"csiso2022jp" => EncodingLabel::Iso2022Jp,
    b"iso-2022-jp" => EncodingLabel::Iso2022Jp,

    // Shift_JIS
    b"csshiftjis" => EncodingLabel::ShiftJis,
    b"ms932" => EncodingLabel::ShiftJis,
    b"ms_kanji" => EncodingLabel::ShiftJis,
    b"shift-jis" => EncodingLabel::ShiftJis,
    b"shift_jis" => EncodingLabel::ShiftJis,
    b"sjis" => EncodingLabel::ShiftJis,
    b"windows-31j" => EncodingLabel::ShiftJis,
    b"x-sjis" => EncodingLabel::ShiftJis,

    // EUC-KR
    b"cseuckr" => EncodingLabel::EucKr,
    b"csksc56011987" => EncodingLabel::EucKr,
    b"euc-kr" => EncodingLabel::EucKr,
    b"iso-ir-149" => EncodingLabel::EucKr,
    b"korean" => EncodingLabel::EucKr,
    b"ks_c_5601-1987" => EncodingLabel::EucKr,
    b"ks_c_5601-1989" => EncodingLabel::EucKr,
    b"ksc5601" => EncodingLabel::EucKr,
    b"ksc_5601" => EncodingLabel::EucKr,
    b"windows-949" => EncodingLabel::EucKr,

    // replacement
    b"csiso2022kr" => EncodingLabel::Replacement,
    b"hz-gb-2312" => EncodingLabel::Replacement,
    b"iso-2022-cn" => EncodingLabel::Replacement,
    b"iso-2022-cn-ext" => EncodingLabel::Replacement,
    b"iso-2022-kr" => EncodingLabel::Replacement,
    b"replacement" => EncodingLabel::Replacement,

    // UTF-16BE
    b"unicodefffe" => EncodingLabel::Utf16Be,
    b"utf-16be" => EncodingLabel::Utf16Be,

    // UTF-16LE
    b"csunicode" => EncodingLabel::Utf16Le,
    b"iso-10646-ucs-2" => EncodingLabel::Utf16Le,
    b"ucs-2" => EncodingLabel::Utf16Le,
    b"unicode" => EncodingLabel::Utf16Le,
    b"unicodefeff" => EncodingLabel::Utf16Le,
    b"utf-16" => EncodingLabel::Utf16Le,
    b"utf-16le" => EncodingLabel::Utf16Le,

    // x-user-defined
    b"x-user-defined" => EncodingLabel::XUserDefined,
    };
}
